### PR TITLE
Node Contexts Tidy

### PR DESCRIPTION
**Pull request by trisyoungs**
As part of work towards #1455 this PR tidies up "context handling" of procedure nodes.  Essentially, #1455 requires some new functionality to be introduced to the complement of `ProcedureNode`s in the form of `if`-style nodes, which are essentially usable in any context. Introducing that, and handling the checks in a nice way, warranted a bit of a refactor.

### CODE REVIEW

File: src/procedure/nodes/node.cpp [hunk 1]
```diff
@@ -74,11 +73,14 @@ EnumOptions<ProcedureNode::NodeContext> ProcedureNode::nodeContexts()
...
+                                                                   {ProcedureNode::OperateContext, "Operate"},
+                                                                   {ProcedureNode::OperateContext, "Any"},
+                                                                   {ProcedureNode::ParentProcedureContext, "ParentProcedure"}});
 }
 
-ProcedureNode::ProcedureNode(ProcedureNode::NodeType nodeType, ProcedureNode::NodeClass classType)
-    : class_(classType), type_(nodeType)
+ProcedureNode::ProcedureNode(ProcedureNode::NodeType nodeType, std::vector<NodeContext> relevantContexts,
+                             ProcedureNode::NodeClass classType)
+    : class_(classType), type_(nodeType), relevantContexts_(std::move(relevantContexts))
```
Comment (jws-1) [inline src/procedure/nodes/node.cpp:83]: I think I'd prefer if the ordering of the parameters matched the ordering of the member initialisation, although I assume there is a reason why it doesn't!

Comment (jws-1) [inline src/procedure/nodes/node.cpp:83]: I think I'd prefer if the ordering of the parameters matched the ordering of the member initialisations, but I'm sure there is a good reason that it doesn't!

**Review by jws-1 — APPROVED** [review]
I like it - the changes seem reasonable to me, from what I can understand! The negative diff is nice to see, also. I had one stylistic comment/question, but ultimately it's up to you what you do with it!

File: src/procedure/nodes/node.cpp (hunk 1 above)
Comment (trisyoungs) [reply src/procedure/nodes/node.cpp:83]: Nope - just the order the members were defined in the header.  However, I agree that they should reflect the ctor ordering, so have reordered things accordingly.

**Pull request merged**